azure-functions/src/lib.rs

```diff
@@ -213,31 +213,55 @@ fn create_host_file(script_root: &Path, verbose: bool) {
     .unwrap_or_else(|_| panic!("Failed to create '{}'", host_json.display()));
 }
 
+/*
+ * Merge implementation with Serde, from:
+ * https://stackoverflow.com/a/47142105
+ */
+fn merge(a: &mut Value, b: Value) {
+    match (a, b) {
+        (a @ &mut Value::Object(_), Value::Object(b)) => {
+            let a = a.as_object_mut().unwrap();
+            for (k, v) in b {
+                merge(a.entry(k).or_insert(Value::Null), v);
+            }
+        }
+        (a, b) => *a = b,
+    }
+}
+
 fn create_local_settings_file(script_root: &Path, worker_dir: &Path, verbose: bool) {
     let settings = script_root.join("local.settings.json");
+    let mut local: Value;
+    let required_json = json!({
+        "Values" : {
+            "FUNCTIONS_WORKER_RUNTIME": "Rust",
+            "languageWorkers:workersDirectory": worker_dir.parent().unwrap()
+        }
+    });
+    /*
+     * If the settings exists, we need to insert the right workersDirectory setting
+     */
     if settings.exists() {
```

Comment:
So this appears to merge in the values with a settings file that might already exist in the script root, but makes no use of the `local.settings.json` file that was generated from `cargo func new`.

However, the problem is finding the `local.settings.json`.  I guess we could check for one in the current working directory, so if you run `cargo run -- init` or `cargo func run` in the same directory where it exists, we'd copy it over.

We could also use `env!("CARGO_MANIFEST_DIR")`, which should be the directory with `Cargo.toml` and thus `local.settings.json`.  However, that would only be set during `cargo run`; if users manually executed the executable and ran the `init` command, it wouldn't be set.  At which point we could probably fallback to the current directory.

What's your take on this?

Reply:
> So this appears to merge in the values with a settings file that might already exist in the script root, but makes no use of the local.settings.json file that was generated from cargo func new-app. 

With the exception of the "optionally encrypting it", I'm not sure I'm understanding the differences you're considering here. As far as I can tell a `cargo func run` does a `--sync` which brings over the `local.settings.json` should it exist, and then the code in this pull request will merge the `required_json` nonetheless.


I could be misunderstanding the problem, in which case, I have no opinion :smile_cat:

Reply:
`--sync` is independent of this as it controls syncing the Azure Functions Host extensions, which involves generating a .NET project and restoring NuGet packages.  Other than files it puts in `bin/` in the script root, it doesn't affect the script root in any other way.

What I mean is that I think the intent of having `cargo func new-app` generate `local.settings.json` into a new project is to allow users to maintain the settings of their application in that file; it should get copied to the script root during initialization.

However, I'm not sure how we would merge values for an encrypted `local.settings.json` anyway, since the encryption isn't done per-value (i.e. it'd be nice to say "this value isn't encrypted" somehow, since we really only care about the language and workers path).

Does that make sense?

Reply:
I believe I understand, but I haven't any idea of the "right thing to do" here :slightly_frowning_face:

Reply:
No worries!  Mind if I push up a commit to the PR of what I have in mind later tonight and you can review it?

Reply:
Sure thing boss

Reply:
Also, feel free to push commits to this branch, or incorporate commits from this PR.

Reply:
Sorry for the delay.  I pushed up a few new commits that I hope will make sense.

When you get a moment, would you mind reviewing the new commits?

Reply:
Also, once this gets merged, I think I'll refactor the worker commands implementation to look a little bit more like the SDK (better output, discrete command implementations) so that there isn't this jumble of methods in `lib.rs`.

Reply:
One last thing: I discovered that the host respects a relative path for the workers directory (from CWD, which will be the script root), so we no longer need to merge any settings! 🎉  If users delete the settings, the host will error, but I'm not particularly worried about that use case.